lib/application.js

```diff
@@ -241,6 +244,66 @@ app.use = function use(fn) {
 };
 
 /**
+ * Proxy `Router#disuse()` to remove middleware from the app router.
+ * See Router#disuse() documentation for details.
+ *
+ * If the _fn_ parameter is an express app, then it will be
+ * dismounted from the _route_ specified.
+ *
+ * @public
+ */
+
+app.disuse = function use(fn) {
+  var offset = 0;
+  var path = '/';
+
+  // default path to '/'
+  // disambiguate app.use([fn])
+  if (typeof fn !== 'function')  {
+    var arg = fn;
+
+    while (Array.isArray(arg) && arg.length !== 0)  {
+      arg = arg[0];
+    }
+
+    // first arg is the path
+    if (typeof arg !== 'function')  {
+      offset = 1;
+      path = fn;
+    }
+  }
+
+  var fns = flatten(slice.call(arguments, offset));
+
+  if (fns.length === 0)  {
+    throw new TypeError('app.use() requires middleware functions');
+  }
+
+  // setup router
+  this.lazyrouter();
+  var router = this._router;
+
+  fns.forEach(function (fn)  {
+    // non-express app
+    if (!fn || !fn.handle || !fn.set)  {
+      return router.disuse(path, fn);
+    }
+
+    debug('.disuse app under %s', path);
```

Comment:
Hey Douglas, as afar as I can tell during the app mounting using
app.use(anotherApp), the only things that seem to be set up are the parent
and mountpath properties, which I do remove with disuse . To make sure that
we use the same function for the disuse with the router, I've added a
property called mountedRouter on the subapp which holds the middleware
entry point , which I use to call disuse on the app's router  , aside from
that I don't see any other setup that occurs when mounting a subapp, but
please correct me if I'm wrong. I'm not using the disuse function as a
destructor on the subapp if that's what you're asking :) , just removing it
from the routing stack

I'll add  a PR to the npm router when we think this is in a usable state :)

I can see the use of the middleware indexing via the idx parameter in the
routers.handle function. To circumvent problems with router indexing and
callback coordination, I've decided that the current best way of handling
the disuse of middleware in the router is to null the value of the stack
and perform a check  via (!layer) to determine whether to continue in the
stack . There are alternate ways to do it , but I think that the
performance hit on servers which have high traffic will be too severe
attempting to  coordinate all the currently running requests , so I've
decided against it, but I do think that the stack would need to be cleaned
at a later stage to stop the stack size from growing too large as a
possibilitiy

On Wed, Jul 1, 2015 at 1:17 AM, Douglas Christopher Wilson <
notifications@github.com> wrote:

> In lib/application.js
> <https://github.com/strongloop/express/pull/2695#discussion_r33634834>:
>
> > +
> > +  if (fns.length === 0)  {
> > +    throw new TypeError('app.use() requires middleware functions');
> > +  }
> > +
> > +  // setup router
> > +  this.lazyrouter();
> > +  var router = this._router;
> > +
> > +  fns.forEach(function (fn)  {
> > +    // non-express app
> > +    if (!fn || !fn.handle || !fn.set)  {
> > +      return router.disuse(path, fn);
> > +    }
> > +
> > +    debug('.disuse app under %s', path);
>
> So when you use an app, we do a bunch of manipulation on it like setting
> up setting inheritance, req/res protos, etc. I'm not sure I see that being
> undone here. If not, please add it :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/strongloop/express/pull/2695/files#r33634834>.
>